### PR TITLE
Adding persistance

### PR DESCRIPTION
**Pull request by jhb15**
Got all code in place for getting/writing/updating/deleting Token to UserId mappings. Also some code in place for getting/writing/updating/deleting AppId and AppSecret. I have added tests to test this code which should pass locally if a DB connection is setup but I had to modify them so they would not run(skip) if it can't connect to a database.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -8,5 +8,71 @@
...
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-entitymanager</artifactId>
+            <version>5.3.7.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>42.2.5.jre7</version>
+        </dependency>
+        <!--<dependency>
```
Comment (JackThomson2) [inline pom.xml]: I presume we don't need this dependency?

Comment (jhb15) [reply pom.xml]: Oh yeah I only needed it for the local DB i had setup.

**Comment by JackThomson2** [comment]
It also looks like Travis is skipping all the tests.

File: fitbit-ingest-service.iml [hunk 2]
```diff
@@ -12,15 +12,61 @@
...
+    <orderEntry type="library" name="Maven: net.bytebuddy:byte-buddy:1.8.17" level="project" />
+    <orderEntry type="library" name="Maven: antlr:antlr:2.7.7" level="project" />
+    <orderEntry type="library" name="Maven: org.jboss.spec.javax.transaction:jboss-transaction-api_1.2_spec:1.1.1.Final" level="project" />
+    <orderEntry type="library" name="Maven: org.jboss:jandex:2.0.5.Final" level="project" />
+    <orderEntry type="library" name="Maven: com.fasterxml:classmate:1.3.4" level="project" />
+    <orderEntry type="library" name="Maven: javax.activation:javax.activation-api:1.2.0" level="project" />
+    <orderEntry type="library" name="Maven: org.dom4j:dom4j:2.1.1" level="project" />
+    <orderEntry type="library" name="Maven: org.hibernate.common:hibernate-commons-annotations:5.0.4.Final" level="project" />
+    <orderEntry type="library" name="Maven: org.hibernate:hibernate-entitymanager:5.3.7.Final" level="project" />
+    <orderEntry type="library" name="Maven: org.postgresql:postgresql:42.2.5.jre7" level="project" />
```
Comment (dkmonaghan) [inline fitbit-ingest-service.iml]: Update your branch without your testing dependencies - pull request will automatically update itself once you update the adding_persistence branch.

File: pom.xml [hunk 3]
```diff
@@ -8,5 +8,71 @@
...
+            <artifactId>hibernate-core</artifactId>
+            <version>5.3.7.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-entitymanager</artifactId>
+            <version>5.3.7.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
```
Comment (dkmonaghan) [inline pom.xml]: As above.

File: src/main/java/persistance/ClientCredentials.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package persistance;
+
+import javax.persistence.*;
+
+@Entity(name = "client_credentials")
+public class ClientCredentials {
```
Comment (dkmonaghan) [inline src/main/java/persistance/ClientCredentials.java]: All of this class is missing Javadoc comments - please could you add them

File: src/main/java/persistance/StorageManager.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            return tokenMap.getAccessToken();
+        }
+    }
+
+    /**
+     * This function removes a record from persistent storage.
+     * @param uId user id of the current user
+     * @return 0 for success, -1 failure (record doesn't exist)
+     */
+    public int removeTokenMap(String uId) {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Why does this return an integer to determine failure / success? Can we not use a true/false bool here?

File: src/main/java/persistance/StorageManager.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        TokenMap tokeMap;
+        try {
+            tokeMap = getTokenMap(uId);
+        } catch (NoResultException nre) {
+            tokeMap = null;
+        }
+        return tokeMap;
+    }
+
+    private void createTokenMap(String uID, String accessToken, String refreshToken) {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Missing Javadoc comments

File: src/main/java/persistance/StorageManager.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            tokenMap.setRefreshToken(refreshToken);
+            em.getTransaction().begin();
+            em.persist(tokenMap);
+            em.getTransaction().commit();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+    
+    private void updateTokenMap(TokenMap tm, String accessToken, String refreshToken) {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Missing Javadoc comments

File: src/main/java/persistance/StorageManager.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        TokenMap tokenMap = em.find(TokenMap.class, tm.getId());
+        em.getTransaction().begin();
+        tm.setAccessToken(accessToken);
+        tm.setRefreshToken(refreshToken);
+        em.getTransaction().commit();
+    }
+
+    private TokenMap getTokenMap(String uId) {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Missing Javadoc comments

File: src/main/java/persistance/StorageManager.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        if (creds == null)
+            return null;
+        return creds.getClientSecret();
+    }
+
+    /**
+     * This funtion removes Application Credentials
+     * @return
+     */
+    public int removeAppCreds() {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Again, could this be a boolean? Javadoc comment also needs finishing off.

File: src/main/java/persistance/StorageManager.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        ClientCredentials clientCredentials;
+        try {
+            clientCredentials = getClientCredentials();
+        } catch (NoResultException nre) {
+            clientCredentials = null;
+        }
+        return clientCredentials;
+    }
+
+    private void  updateCredRecord(ClientCredentials cred, String id, String secret) {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Missing Javadoc comments

File: src/main/java/persistance/StorageManager.java [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        em.getTransaction().begin();
+        cred.setClientId(id);
+        cred.setClientSecret(secret);
+        em.getTransaction().commit();
+    }
+
+    private void createCredRecord(String id, String secret) {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Missing Javadoc comments

File: src/main/java/persistance/StorageManager.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        ClientCredentials creds = new ClientCredentials();
+        creds.setClientId(id);
+        creds.setClientSecret(secret);
+        creds.setService("fitbit");
+        em.getTransaction().begin();
+        em.persist(creds);
+        em.getTransaction().commit();
+    }
+
+    private ClientCredentials getClientCredentials() {
```
Comment (dkmonaghan) [inline src/main/java/persistance/StorageManager.java]: Missing Javadoc comments

File: src/main/java/persistance/TokenMap.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
+package persistance;
+
+import javax.persistence.*;
+import java.io.Serializable;
+
+@Entity(name = "token_map")
+public class TokenMap implements Serializable {
```
Comment (dkmonaghan) [inline src/main/java/persistance/TokenMap.java]: This class is missing Javadoc comments throughout.

File: src/main/java/persistance/readme.txt [hunk 14]
```diff
@@ -0,0 +1,17 @@
+How to Connect Local DB:
```
Comment (dkmonaghan) [inline src/main/java/persistance/readme.txt]: This information is probably going to get overlooked if this readme.txt is in this directory. I'd recommend you make a 'docs' folder in the root of your repository and put this file there. 

(also, give it a .md file extension so GitHub renders it nicely)

File: src/main/resources/META-INF/persistence.xml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+<persistence xmlns="http://xmlns.jcp.org/xml/ns/persistence"
+             xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="2.2"
+             xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/persistence http://xmlns.jcp.org/xml/ns/persistence/persistence_2_2.xsd">
+
+    <persistence-unit name="fitbitPU">
+        <provider>org.hibernate.jpa.HibernatePersistenceProvider</provider>
+        <class>persistance.TokenMap</class>
+        <class>persistance.ClientCredentials</class>
+        <properties>
+            <property name="hibernate.connection.url" value="jdbc:postgresql://localhost:5432/fitbit_digest_db"/>
```
Comment (dkmonaghan) [inline src/main/resources/META-INF/persistence.xml]: Hardcoded PostgreSQL connection information - probably needs removing too?

**Comment by dkmonaghan** [comment]
> It also looks like Travis is skipping all the tests. Not sure how we want to tackle that.

@DavidFair any ideas?

**Comment by DavidFair** [comment]
The tests were skipped:
> Tests run: 6, Failures: 0, Errors: 0, Skipped: 6

I suspect it is related to these warnings, but I would need to put a breakpoint on a test to see if they are even entered:
```
Oct 30, 2018 10:28:50 PM org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl configure
WARN: HHH10001002: Using Hibernate built-in connection pool (not for production use!)
Oct 30, 2018 10:28:50 PM org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl buildCreator
INFO: HHH10001005: using driver [org.postgresql.Driver] at URL [jdbc:postgresql://localhost:5432/fitbit_digest_db]
Oct 30, 2018 10:28:50 PM org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl buildCreator
INFO: HHH10001001: Connection properties: {user=james, password=****}
Oct 30, 2018 10:28:50 PM org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl buildCreator
INFO: HHH10001003: Autocommit mode: false
Oct 30, 2018 10:28:50 PM org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl$PooledConnections <init>
INFO: HHH000115: Hibernate connection pool size: 20 (min=1)
Oct 30, 2018 10:28:50 PM org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator initiateService
WARN: HHH000342: Could not obtain connection to query metadata : null
```

**Comment by jhb15** [comment]
I have now added the JavaDoc comments and also removed the -1, 0 return messages. I also removed that hard-coded DB properties and left them empt for now as that didn't cause a build failure. I also removed the PostgreSQL driver dependency I was using for testing.

File: fitbit-ingest-service.iml (hunk 2 above)
Comment (jhb15) [reply fitbit-ingest-service.iml]: removed the postgresql testing dependency but the hibernate stuff isn't for testing its required.

File: pom.xml (hunk 3 above)
Comment (jhb15) [reply pom.xml]: done

File: src/main/java/persistance/ClientCredentials.java (hunk 4 above)
Comment (jhb15) [reply src/main/java/persistance/ClientCredentials.java]: done

File: src/main/java/persistance/StorageManager.java (hunk 5 above)
Comment (jhb15) [reply src/main/java/persistance/StorageManager.java]: done

File: src/main/java/persistance/StorageManager.java (hunk 6 above)
Comment (jhb15) [reply src/main/java/persistance/StorageManager.java]: done

File: src/main/java/persistance/StorageManager.java (hunk 11 above)
Comment (jhb15) [reply src/main/java/persistance/StorageManager.java]: done

**Pull request merged**